datafusion/optimizer/src/handle_gap_fill.rs

```diff
@@ -0,0 +1,451 @@
+    }
+
+    let mut new_aggr_expr = Vec::with_capacity(group_expr.len());
+    let mut dbg_idx = None;
+
+    group_expr
+        .iter()
+        .enumerate()
+        .try_for_each(|(i, e)| -> Result<()> {
+            let new_expr = match e {
+                Expr::ScalarFunction {
+                    fun: DateBinGapfill,
+                    args,
+                } => {
+                    if dbg_idx.is_some() {
+                        return Err(DataFusionError::Plan(format!(
+                            "DATE_BIN_GAPFILL specified more than once"
+                        )));
+                    }
+                    dbg_idx = Some(i);
+                    Expr::ScalarFunction {
+                        fun: DateBin,
+                        args: args.clone(),
+                    }
+                }
+                _ => e.clone(),
+            };
+            new_aggr_expr.push(new_expr);
+            Ok(())
+        })?;
+    Ok(Some((new_aggr_expr, dbg_idx.unwrap())))
+}
+
+fn collect_gap_fns(e: &Expr) -> Result<Vec<BuiltinScalarFunction>> {
+    struct Finder {
+        fns: Vec<BuiltinScalarFunction>,
+    }
+    impl ExpressionVisitor for Finder {
+        fn pre_visit(mut self, expr: &Expr) -> Result<Recursion<Self>> {
+            if let Expr::ScalarFunction {
+                fun:
+                    f @ (BuiltinScalarFunction::DateBinGapfill | BuiltinScalarFunction::LOCF),
+                ..
+            } = expr
+            {
+                self.fns.push(f.clone())
+            };
+            Ok(Recursion::Continue(self))
+        }
+    }
+    let f = Finder { fns: vec![] };
+    let f = e.accept(f)?;
+    Ok(f.fns)
+}
+
+fn check_node(node: &LogicalPlan) -> Result<()> {
+    println!("checking node");
+    node.expressions().iter().try_for_each(|expr| {
+        let fns = collect_gap_fns(expr)?;
+        if !fns.is_empty() {
+            Err(DataFusionError::Plan(format!(
+                "invalid context for {}",
+                fns[0]
+            )))
+        } else {
+            Ok(())
+        }
+    })?;
+    Ok(())
+}
+
+/// A logical node that represents the gap filling operation.
+///
+/// This was made a user-defined node for easier iterative development
+/// but it's representation may change. The inner plan node is always
+/// an Aggregate node, since it needs similar inputs.
+#[derive(Clone, Debug)]
+struct GapFill {
+    inner: LogicalPlan,
+}
+
+impl UserDefinedLogicalNode for GapFill {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn inputs(&self) -> Vec<&LogicalPlan> {
+        let input = Aggregate::try_from_plan(&self.inner)
+            .expect("this should always be an Aggregate node")
+            .input
+            .as_ref();
+        vec![input]
+    }
+
+    fn schema(&self) -> &datafusion_common::DFSchemaRef {
+        &Aggregate::try_from_plan(&self.inner)
+            .expect("this should aways be an Aggregate node")
+            .schema
```

Comment:
You can also just call schema on the input I think

```suggestion
        self.inner
            .schema()
```